src/fetching.rs

```diff
@@ -0,0 +1,135 @@
+//! type of resource from a specific source. The [`provider`] module contains built-in
+//! implementations of [`Provider`] for various data availability sources.
+//!
+
+use async_std::{
+    sync::{Arc, Mutex},
+    task::spawn,
+};
+use derivative::Derivative;
+use std::collections::{hash_map::Entry, BTreeSet, HashMap};
+
+pub mod provider;
+pub mod request;
+
+pub use provider::Provider;
+pub use request::Request;
+
+/// A callback to process the result of a request.
+///
+/// Sometimes, we may fetch the same object for multiple purposes, so a request may have more than
+/// one callback registered. For example, we may fetch a leaf for its own sake and also to
+/// reconstruct a block. Or, we may fetch the same payload for two different blocks. In both of
+/// these cases, there are two objects that must be processed and stored after the fetch completes.
+///
+/// In these cases, we only want one task to actually fetch the resource, but there may be several
+/// unrelated actions to take after the resource is fetched. This trait allows us to identify a
+/// callback, so that when the task that actually fetched the resource completes, it will run one
+/// instance of each distinct callback which was registered. Callbacks will run in the order
+/// determined by `Ord`.
+#[trait_variant::make(Callback: Send)]
+pub trait LocalCallback<T>: Ord {
+    async fn run(self, response: T);
+}
+
+/// Management of concurrent requests to fetch resources.
+#[derive(Derivative)]
+#[derivative(Clone(bound = ""), Debug(bound = ""), Default(bound = ""))]
+pub struct Fetcher<T, C> {
+    #[derivative(Debug = "ignore")]
+    in_progress: Arc<Mutex<HashMap<T, BTreeSet<C>>>>,
+}
+
+impl<T, C> Fetcher<T, C> {
+    /// Fetch a resource, if it is not already being fetched.
+    ///
+    /// This function will spawn a new task to fetch the resource in the background, using callbacks
+    /// to process the fetched resource upon success. If the resource is already being fetched, the
+    /// spawned task will terminate without fetching the resource, but only after registering the
+    /// provided callbacks to be executed by the existing fetching task upon completion, as long as
+    /// there are not already equivalent callbacks registered.
+    ///
+    /// We spawn a (short-lived) task even if the resource is already being fetched, because the
+    /// check that the resource is being fetched requires an exclusive lock, and we do not want to
+    /// block the caller, which might be on the critical path of request handling.
+    ///
+    /// Note that while callbacks are allowed to be async, they are executed sequentially while an
+    /// exclusive lock is held, and thus they should not take too long to run or block indefinitely.
+    pub fn spawn_fetch<Types>(
+        &self,
+        req: T,
+        provider: impl Provider<Types, T> + 'static,
+        callbacks: impl IntoIterator<Item = C> + Send + 'static,
```

Comment:
Does it make sense to somehow limit the number of callbacks?

Reply:
I don't think it's necessary. Callbacks are deduplicated, plus the list of callbacks is determined entirely by our own internal logic, it's not subject to user input